### ISSUE TITLE
Create a Dark Theme

### ISSUE THREAD

**Issue by adamdbradley** [driftyco/ionic#4967]
_From @adamdbradley on January 11, 2016 14:54_

Ionic comes with a default theme, which is mainly a white background and black foreground:
https://github.com/driftyco/ionic2/blob/master/ionic/themes/default.scss

However, Ionic2's folder structure, Sass imports and variables have all been setup to allow other themes to be easily added, and allow end developers to choose which theme they want to start from. For example, an app could start with the dark theme, but they could still customize all the colors to fit their brand.

If anyone is looking to help create the dark theme we'd really appreciate the assistance. A lot of the work is just making sure all of the css and sass variables are in place to allow the entire app's theme to switch out, and allow many other themes to be easily created with minimal sass variable changes. If interested please let us know, thanks!

_Copied from original issue: driftyco/ionic2#923_

**Comment by adamdbradley** [comment]
_From @alexbainbridge on January 12, 2016 0:20_

I know you know this, but v2 docs suggests the dark theme already exists http://ionicframework.com/docs/v2/theming/theming-your-app/ 

I tried to use it then found it was docs only :)

**Comment by adamdbradley** [comment]
Yeah sorry, it did at one point. But as we were building our the sass files it got to be a pain to keep updating. But now that things are solidifying we can start building it out again.

**Comment by adamdbradley** [comment]
_From @matheusrocha89 on January 12, 2016 11:6_

I wanna help you guys with this :+1:

**Comment by adamdbradley** [comment]
@matheusrocha89 Awesome thanks! I'll assign it to myself so for now so know it's being worked on. Any questions please let us know and we can help make sure the files are setup correctly. The new `dark.scss` file should go in here: https://github.com/driftyco/ionic2/tree/master/ionic/themes

Thanks!

**Comment by adamdbradley** [comment]
_From @jgw96 on January 13, 2016 18:53_

@matheusrocha89 are you working on this? Id love to help if needed?

**Comment by adamdbradley** [comment]
_From @matheusrocha89 on January 13, 2016 20:59_

@jgw96 I will start to do something today, I will create a branch. Any help it's welcome :+1:

**Comment by adamdbradley** [comment]
Heads up, we JUST moved `ionic2` to be in a different branch in the `ionic` repo: https://github.com/driftyco/ionic/tree/2.0

**Comment by adamdbradley** [comment]
_From @jgw96 on January 13, 2016 22:50_

@adamdbradley yep saw that today. Are all these bugs moved over to the main repo too?

**Comment by adamdbradley** [comment]
_From @jgw96 on January 13, 2016 22:50_

@adamdbradley or are all the bugs still gonna be handled in this repo?

**Comment by adamdbradley** [comment]
Currently being moved over, we've got some work to do to clean up the process for creating and managing issues.

**Comment by adamdbradley** [comment]
_From @jgw96 on January 13, 2016 22:52_

Got ya, is there gonna be an official set of contributing guidelines set up for v2?

**Comment by matheusrocha89** [comment]
@adamdbradley, @jgw96 made a good point, is there any contributing guidelines?

**Comment by 3dd13** [comment]
there are several contributing guideline documents for ionic-related repos.

I usually follow the first link below to build and test changes:
* https://github.com/driftyco/ionic/blob/2.0/scripts/README.md
* https://github.com/driftyco/ionic/blob/af1bfef327e685585244c6051c4d38b98aa6c62a/CONTRIBUTING.md
* http://ionicframework.com/contribute/#pull-request-guidelines
* https://github.com/driftyco/ionic-conference-app/blob/2bd80de4874f3f363cf3717f0b986fb5aa85c727/CONTRIBUTING.md

**Comment by 3dd13** [comment]
for look and feel changes, perhaps should suggest contributors to build a demo page or update ionic-site repo as well ?

**Comment by adamdbradley** [comment]
@jgw96 @3dd13 @matheusrocha89 Just checking in to see if there are any questions regarding this. Also, we have a slack channel for ionic2 development, if you're interested please email, adam at ionic io

**Comment by jgw96** [comment]
Don't have any questions yet on my end, I plan on working on this tonight,
I'd love to be a part of that slack channel! Shooting you an email now.

**Comment by jgw96** [comment]
@adambradley is there any specific style guides that the dark theme should follow? Such as a specific color pallete? My thinking on this is that the $dark color in the normal theme should be the basis of the dark theme.

**Comment by matheusrocha89** [comment]
@adamdbradley If the slack is Ionic Worldwide I am already there :+1: Are you there too?

**Comment by matheusrocha89** [comment]
@adamdbradley I'm trying to create the dark theme and test the scss file in an example project to see how it's going on, but on the documentation it's says that I just need to import the "theme" on `app.scss` file and this file is importing ionic.scss but I can't find this app.scss file, I just found app.core.scss that it's not importing the ionic scss

**Comment by matheusrocha89** [comment]
@jgw96 Are you on Ionic Worldwide slack?

**Comment by matheusrocha89** [comment]
@adamdbradley I already create my dev enviroment to test the themes. I just have one question, do you have any idea of how do you want the dark theme? All dark, just toolbars dark? Or am I free to do whatever I want?

**Comment by adamdbradley** [comment]
It should be the entire app, like these examples:
![image](https://cloud.githubusercontent.com/assets/452425/12511926/a64f85e8-c0da-11e5-99c9-2f845eb9d4ad.png)

![image](https://cloud.githubusercontent.com/assets/452425/12511958/e520e80c-c0da-11e5-9604-075b4b3a8bb3.png)

Also, the spotify app is a good example of a dark themed app:

![image](https://cloud.githubusercontent.com/assets/452425/12511987/0fb85244-c0db-11e5-83d1-217311e4368a.png)

For now you can guess some good colors to use, but we'll probably end up tweaking them a bit once @bensperry gets ahold of them. Any other questions please let me know, thanks

**Comment by matheusrocha89** [comment]
@adamdbradley :+1:

**Comment by brandyscarney** [comment]
The links to the code are broken currently, but there is a searchable list of all of the Sass variables with their default values that may be helpful: http://ionicframework.com/docs/v2/theming/overriding-ionic-variables/

Subscribing to this issue to also answer any questions. :smile:

**Comment by matheusrocha89** [comment]
@brandyscarney I used this to run the ionic-site example https://github.com/driftyco/ionic/blob/2.0/scripts/README.md on my mac. Is there a way to force the v2 docs of ionic-site example to use the new theming that I am creating? I would like to edit my theme and see how it's looking with the docs examples

**Comment by brandyscarney** [comment]
@matheusrocha89 We moved all of the demos to their own repo - https://github.com/driftyco/ionic-preview-app (I just made it public). So you could import the dark theme in this app and then run it just like a normal Ionic 2 app.

**Comment by adamdbradley** [comment]
@matheusrocha89 I made some updates to the repo to make it easier to create new themes. If you get the latest from the `2.0` branch, I made a few changes to the theme structures, and added the dark theme scss files: https://github.com/driftyco/ionic/tree/2.0/ionic/themes

What I have there now is pretty much the same as the default theme, but with a black background and white text. If you can tweak the colors a bit to be more like the examples above that'd be great, but as it gets further along @bensperry well be helping to set the final the colors for the dark theme.

To view e2e with the dark theme:

- `gulp sass.themes`
- `gulp watch`
- Default theme e2e tests: <http://localhost:8000/dist/e2e/>
- Put `?theme=dark` at the end of any e2e test to see it with the dark theme

When you make changes to the dark themes you'll have to run `gulp sass.themes` again to rebuild the e2e css files.

So the next steps are to come up with some good dark colors, then view each of the e2e tests and make sure each component can invert colors, and that there aren't any issues with components not styling correctly. Any questions please let us know. Thanks!

**Comment by adamdbradley** [comment]
@matheusrocha89 Also, I'd prefer this is done in smaller steps, rather than one large PR. If you can break up changes to smaller commits and narrowed updating one component or color change that'd be great, and it'll be easier for us to merge in changes. Thanks

**Comment by adamdbradley** [comment]
![image](https://cloud.githubusercontent.com/assets/452425/12755949/1ef5df46-c997-11e5-8727-11cf58df7d2a.png)

**Comment by bensperry** [comment]
Looking nice!!

**Comment by matheusrocha89** [comment]
Ok, I will try, thanks @brandyscarney and @adamdbradley

**Comment by matheusrocha89** [comment]
@adamdbradley one last question, is there a way to use the md version on e2e too?

**Comment by adamdbradley** [comment]
Either change the device with chrome's dev tools: 
![image](https://cloud.githubusercontent.com/assets/452425/12757808/30b58896-c99f-11e5-8197-df0b661cc153.png)

Or add this querystring `ionicplatform=android`

**Comment by matheusrocha89** [comment]
:+1:

**Comment by matheusrocha89** [comment]
@brandyscarney Hi, I created a version of the dark theme but I am having some problems with the chips component, because they don't have any variables to override and since they are dark the became almost invisible on dark theme. Do you have any idea of how we can find a solution for them?

**Comment by brandyscarney** [comment]
@matheusrocha89 Hey, if you want to ignore the chip component for now I can work on adding some variables in this issue: https://github.com/driftyco/ionic/issues/5386 :smile:

**Comment by tomsun** [comment]
"allow end developers to choose which theme they want to start from"
https://github.com/driftyco/ionic/issues/4967#issue-126730040

How is this accomplished (or what is the plan)? 

(A starter-app currently loads `ionic.ios.scss` which loads `components.ios` which loads `globals.ios` which loads `themes/default.ios`...
Naive guess #1: somehow swap `themes/default.ios` for `themes/dark.ios` in the above chain even though you don't have ownership of those files?
Naive guess #2: `themes/dark.ios` by convention shadows all relevant parts of `themes/default.ios` and can be included directly by the app somehow? )

**Comment by tomsun** [comment]
Importing `themes/dark.ios` **before** `'ionic.ios' from the app's scss does make things darker! (my instinct was to inport the theme as the last step, which does nothing)

app.ios.scss
```
@import 'themes/dark.ios';
@import 'ionic.ios';
```

**Comment by ghenry22** [comment]
the ability to switch themes programatically rather than just at build time would be great as then we could allow users to choose from a list of defined themes in app.

**Comment by mg1075** [comment]
Can't tell for sure what is going on with this issue, since it seems it was never truly resolved.
Anyways...

1. There has been recent work on a dark theme by @pliablepixels here:
https://github.com/pliablepixels/ionic-dark-grey-theme

2. And then the other issue is: how can we have the ability to switch themes programmatically, as asked by @ghenry22?

**Comment by ghenry22** [comment]
@mg1075 Looks like some decent work on a dark theme there.

I do have a solution for programmatically switching themes which works perfectly, the only problem that I have to put it into production or share it is that I can't figure out how to adjust the ionic build process to run an extra step to compile and build the alternate css.

For example:
What I need to figure out is how to have a second copy of variables.scss (ie variables_dark.scss) which has the dark app theme selected and any specific colour overrides etc for the dark theme in it, where the default variables.scss has the default ionic light theme.

In the build process I want it to run the usual build against variables.scss, then run those tasks again against variables_dark.scss and generate a main_dark.css.

Then we have compiled and minified light theme (main.css) and dark theme (main_dark.css) and then we programmatically switch those out.

Sorry it can be hard to explain but basically if someone can help with how to add a custom build step as above so that automatic builds work properly then I will gladly share the rest of my solution for switching themes programmatically to pair up with it.

**Comment by pliablepixels** [comment]
I did a bit of investigation on dynamic switching for my [theme](https://github.com/pliablepixels/ionic-dark-grey-theme). I don't have an answer yet, just half-trodden paths with walls I need to scale. I  am also not sure if I'm diving in deeper than needed and if there is a better path to explore.

I looked at two options:

a) In `variables.scss`, scope your imports like this:
```
.dark_grey_theme { @import 'dark.scss' }
```

While this does scope the import and any classes defined in it correctly, it doesn't really solve the problem of ionic components using default colors for its components. My goal here was basically to have a `dark_grey_theme` and a `default` theme scoped import and wrap the app inside a div that I can dynamically toggle. However, this did not work due to reasons I stated above.

b) Get `ionic build` to compile multiple sass files and output them as 'main_dark.css' and 'main_default.css'. Ionic allows you to modify scripts with your custom scripts, which I got working, but couldn't figure out how to make the sass part invoke two different scripts with two prefixes. I used it to generate either dark or default, but not both

Maybe someone with more expertise here can comment on the right path to explore.
